### PR TITLE
add solana

### PR DESCRIPTION
**Pull request by detherminal**
Added Solana with all functionalities:
- mnemonic generation/restoration
- address derivation
- balance control
- transaction control
- transfering
- rent exemption of Solana
- price fetching
etc. etc.

### CODE REVIEW

File: lib/wallets/crypto_currency/coins/solana.dart [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        throw Exception("Unsupported network: $network");
+    }
+  }
+
+  @override
+  int get minConfirms => 21;
+
+  @override
+  bool validateAddress(String address) {
+    RegExp regex = RegExp(r'^[a-zA-Z0-9]{44}$');
```
Comment (sneurlax) [inline lib/wallets/crypto_currency/coins/solana.dart:42]: TODO better validation

File: lib/wallets/wallet/impl/solana_wallet.dart [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+import 'package:stackwallet/models/node_model.dart';
+import 'package:stackwallet/services/node_service.dart';
+
+class SolanaWallet extends Bip39Wallet<Solana> {
+  SolanaWallet(CryptoCurrencyNetwork network) : super(Solana(network));
+
+  NodeModel? _solNode;
+
+  Future<Ed25519HDKeyPair> _getKeyPair() async {
+    return Ed25519HDKeyPair.fromMnemonic(await getMnemonic(), account: 0, change: 0);
```
Comment (sneurlax) [inline lib/wallets/wallet/impl/solana_wallet.dart:30]: Is this Solana's HD path?

File: lib/wallets/wallet/impl/solana_wallet.dart [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    try {
+      var address = (await _getKeyPair()).address;
+
+      await mainDB.updateOrPutAddresses([
+        Address(
+            walletId: walletId,
+            value: address,
+            publicKey: List<int>.empty(),
+            derivationIndex: 0,
+            derivationPath: null,
```
Comment (sneurlax) [inline lib/wallets/wallet/impl/solana_wallet.dart:66]: Would this restore the same address as other popular Solana wallets?  Is there a standard derivation path or several which we should check?

Comment (detherminal) [reply lib/wallets/wallet/impl/solana_wallet.dart:66]: The ones in the _getKeyPair (So rest of the wallet) are fully compatible and same as popular Solana wallets like Phantom and Solflare.

File: lib/wallets/wallet/impl/solana_wallet.dart [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+          feeAmount = currentFees.medium;
+          break;
+      }
+
+      // Rent exemption of Solana
+      final rpcClient = RpcClient("${getCurrentNode().host}:${getCurrentNode().port}");
+      final accInfo = await rpcClient.getAccountInfo((await _getKeyPair()).address);
+      final minimumRent = await rpcClient.getMinimumBalanceForRentExemption(accInfo.value!.data.toString().length);
+      if (minimumRent > ((await _getCurrentBalanceInLamports()) - txData.amount!.raw.toInt() - feeAmount)) {
+        throw Exception("Insufficient remaining balance for rent exemption, minimum rent: ${minimumRent / pow(10, cryptoCurrency.fractionDigits)}");
```
Comment (sneurlax) [inline lib/wallets/wallet/impl/solana_wallet.dart:111]: What if we have less than the minimum rent amount?  Can we still send?

Comment (detherminal) [reply lib/wallets/wallet/impl/solana_wallet.dart:111]: No, the library and the network doesn't allow.

File: lib/wallets/wallet/impl/solana_wallet.dart [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+  @override
+  Future<void> updateBalance() async {
+    try {
+      var rpcClient = RpcClient("${getCurrentNode().host}:${getCurrentNode().port}");
+      var balance = await rpcClient.getBalance(info.cachedReceivingAddress);
+
+      // Rent exemption of Solana
+      final accInfo = await rpcClient.getAccountInfo((await _getKeyPair()).address);
+      final minimumRent = await rpcClient.getMinimumBalanceForRentExemption(accInfo.value!.data.toString().length);
+      var spendableBalance = balance.value - minimumRent;
```
Comment (sneurlax) [inline lib/wallets/wallet/impl/solana_wallet.dart:228]: What if we want to send more than the minimum rent out of the wallet?

Comment (detherminal) [reply lib/wallets/wallet/impl/solana_wallet.dart:228]: The account have to be closed to fully send all funds.

File: lib/wallets/wallet/impl/solana_wallet.dart (hunk 4 above)
Comment (sneurlax) [reply lib/wallets/wallet/impl/solana_wallet.dart:111]: I thought you should be able to reclaim your rent eventually, no?

I'm not very educated on Solana tho.

File: lib/wallets/wallet/impl/solana_wallet.dart (hunk 5 above)
Comment (sneurlax) [reply lib/wallets/wallet/impl/solana_wallet.dart:228]: Can we actually do that?

**Review by sneurlax — APPROVED** [review]
Looks good to me although I still need to review if Tor usage works with the RpcClient as used here

**Comment by sneurlax** [comment]
@detherminal I merged your changes to the `solana` branch, where I will hide Solana behind a toggle in the developer menu so we can test it more thoroughly.  To access the developer menu, click the Stack Wallet logo 7 times (it should spin when clicked)

The notes here on this PR should be followed up on but we might convert them to issues and leave them to be upon improved later

**Pull request merged**